File: examples/template.py
from robodbg import Debugger, BreakpointAction

class PyDebugger(Debugger):
    def __init__(self):
        super().__init__()
        print("[PyDebugger] Initialized")

    def on_start(self, image_base, entry_point):
        print(f"[on_start] Image Base: {hex(image_base)}, Entry Point: {hex(entry_point)}")

    def on_end(self, exit_code, pid):
        print(f"[on_end] Exit Code: {exit_code}, PID: {pid}")

    def on_attach(self):
        print("[on_attach] Attached to process")

    def on_thread_create(self, h_thread, thread_id, thread_base, start_address):
        print(f"[on_thread_create] Thread ID: {thread_id}, Base: {hex(thread_base)}, Start: {hex(start_address)}")

    def on_thread_exit(self, thread_id):
        print(f"[on_thread_exit] Thread ID: {thread_id}")

    def on_dll_load(self, address, dll_name, entry_point):
        print(f"[on_dll_load] {dll_name} at {hex(address)} with EP: {hex(entry_point)}")
        return True

    def on_dll_unload(self, address, dll_name):
        print(f"[on_dll_unload] {dll_name} from {hex(address)}")

    def on_breakpoint(self, address, h_thread):
        print(f"[on_breakpoint] Breakpoint hit at {hex(address)}")
        return BreakpointAction.BREAK

    def on_hardware_breakpoint(self, address, h_thread, reg):
        print(f"[on_hardware_breakpoint] HWBP at {hex(address)} in {reg}")
        return BreakpointAction.BREAK

    def on_single_step(self, address, h_thread):
        print(f"[on_single_step] Address: {hex(address)}")

    def on_debug_string(self, dbg_string):
        print(f"[on_debug_string] {dbg_string}")

    def on_access_violation(self, address, faulting_address, access_type):
        print(f"[on_access_violation] At {hex(address)} -> Faulting {hex(faulting_address)}, Type: {access_type}")

    def on_rip_error(self, rip):
        print(f"[on_rip_error] {rip}")

    def on_unknown_exception(self, addr, code):
        print(f"[on_unknown_exception] At {hex(addr)}, Code: {hex(code)}")

    def on_unknown_debug_event(self, code):
        print(f"[on_unknown_debug_event] Code: {hex(code)}")

File: examples/template.cpp
#pragma once
#include <robodbg/debugger.h>
#include <robodbg/util.h>
#include <iostream>

using namespace RoboDBG;

class MyDebugger : public Debugger {
public:
    MyDebugger() = default;

    // Called when the process starts
    virtual void onStart(uintptr_t imageBase, uintptr_t entryPoint) override {
        std::cout << "[onStart] imageBase: " << std::hex << imageBase
        << ", entryPoint: " << entryPoint << std::endl;
    }

    // Called when the process exits
    virtual void onEnd(DWORD exitCode, DWORD pid) override {
        std::cout << "[onEnd] exitCode: " << std::dec << exitCode
        << ", pid: " << pid << std::endl;
    }

    // Called when attach succeeds
    virtual void onAttach() override {
        std::cout << "[onAttach] Successfully attached to process." << std::endl;
    }

    // New thread created
    virtual void onThreadCreate(HANDLE hThread, DWORD threadId, uintptr_t threadBase, uintptr_t startAddress) override {
        std::cout << "[onThreadCreate] TID: " << threadId << ", base: " << std::hex << threadBase
        << ", start: " << startAddress << std::endl;
    }

    // Thread exited
    virtual void onThreadExit(DWORD threadID) override {
        std::cout << "[onThreadExit] TID: " << threadID << std::endl;
    }

    // DLL loaded
    virtual bool onDLLLoad(uintptr_t address, std::string dllName, uintptr_t entryPoint) override {
        std::cout << "[onDLLLoad] " << dllName << " at " << std::hex << address
        << ", entryPoint: " << entryPoint << std::endl;
        return true;
    }

    // DLL unloaded
    virtual void onDLLUnload(uintptr_t address, std::string dllName) override {
        std::cout << "[onDLLUnload] " << dllName << " from " << std::hex << address << std::endl;
    }

    // Software breakpoint hit
    virtual BreakpointAction onBreakpoint(uintptr_t address, HANDLE hThread) override {
        std::cout << "[onBreakpoint] BP at " << std::hex << address << std::endl;
        return BreakpointAction::BREAK;
    }

    // Hardware breakpoint hit
    virtual BreakpointAction onHardwareBreakpoint(uintptr_t address, HANDLE hThread, DRReg reg) override {
        std::cout << "[onHardwareBreakpoint] DR" << static_cast<int>(reg)
                  << " hit at " << std::hex << address << std::endl;
        return BreakpointAction::BREAK;
    }

    // Single-step trap
    virtual void onSinglestep(uintptr_t address, HANDLE hThread) override {
        std::cout << "[onSinglestep] " << std::hex << address << std::endl;
    }

    // Debug strings (OutputDebugString)
    virtual void onDebugString(std::string dbgString) override {
        std::cout << "[onDebugString] " << dbgString << std::endl;
    }

    // Access violation handler
    virtual void onAccessViolation(uintptr_t address, uintptr_t faultingAddress, long accessType) override {
        std::cout << "[onAccessViolation] " << std::hex << address << " -> "
                  << faultingAddress << ", accessType: " << std::dec << accessType << std::endl;
    }

    // RIP errors (x64 structured exception)
    virtual void onRIPError(const RIP_INFO& rip) override {
        std::cout << "[onRIPError] RIP code: " << std::hex << rip.ErrorCode << std::endl;
    }

    // Unknown exceptions
    virtual void onUnknownException(uintptr_t addr, DWORD code) override {
        std::cout << "[onUnknownException] At: " << std::hex << addr << ", code: " << std::hex << code << std::endl;
    }

    // Unknown debug events
    virtual void onUnknownDebugEvent(DWORD code) override {
        std::cout << "[onUnknownDebugEvent] Code: " << std::hex << code << std::endl;
    }
};
